File: src/longest-palindromic-substring/longest-palindromic-substring.ts
// https://leetcode.com/problems/longest-palindromic-substring/
// https://leetcode.com/submissions/detail/1604517121/

function longestPalindrome(s: string): string {
    function expandFromCenter(left: number, right: number) {
        while (left >= 0 && right < s.length && s[left] === s[right]) {
            left--;
            right++;
        }
        return [left + 1, right - 1];
    }

    if (!s) return "";

    let startIndex = 0;
    let endIndex = 0;

    for (let i = 0; i < s.length; i++) {
        // odd-length palindrome
        const [l1, r1] = expandFromCenter(i, i);
        // even-length palindrome
        const [l2, r2] = expandFromCenter(i, i + 1);

        if ((r1 - l1) > (endIndex - startIndex)) {
            startIndex = l1;
            endIndex = r1;
        }
        if ((r2 - l2) > (endIndex - startIndex)) {
            startIndex = l2;
            endIndex = r2;
        }
    }

    return s.substring(startIndex, endIndex + 1);
};

File: src/longest-palindromic-substring/longest-palindromic-substring.py
# https://leetcode.com/problems/longest-palindromic-substring/

class Solution:
    def longestPalindrome(self, s: str) -> str:
        def expand_from_center(left, right):
            while left >= 0 and right < len(s) and s[left] == s[right]:
                left -= 1
                right += 1
            return left + 1, right - 1

        if not s:
            return ""
        
        start_index, end_index = 0, 0
        for i in range(len(s)):
            # odd palindrome
            l1, r1 = expand_from_center(i, i)
            # even palindrome
            l2, r2 = expand_from_center(i, i + 1)
            
            # choose longest palindrome from both expansions
            if (r1 - l1) > (end_index - start_index):
                start_index, end_index = l1, r1
            if (r2 - l2) > (end_index - start_index):
                start_index, end_index = l2, r2

        return s[start_index:end_index+1]